### PR TITLE
Handle /backfill returning no events

### LINKED ISSUE #10123: Failed to backfill from <server> because UnboundLocalError("local variable 'result' referenced before assignment")
### Description

Python error when backfilling fails.

### Steps to reproduce

The final stacktrace is:
```
UnboundLocalError: local variable 'result' referenced before assignment
  File "synapse/handlers/federation.py", line 1194, in try_backfill
    await self.backfill(
  File "synapse/handlers/federation.py", line 959, in backfill
    state = await self._get_state_for_room(
  File "synapse/handlers/federation.py", line 590, in _get_state_for_room
    await self._get_events_and_persist(
  File "synapse/handlers/federation.py", line 1357, in _get_events_and_persist
    await self._auth_and_persist_events(
  File "synapse/handlers/federation.py", line 2104, in _auth_and_persist_events
    await self.persist_events_and_notify(
  File "synapse/handlers/federation.py", line 3066, in persist_events_and_notify
    return result["max_stream_id"]
```

I've seen different errors further along though:
* `SynapseError: 401: Failed to find any key to satisfy VerifyJsonRequest(server=sugar-free-jazz.com, key_ids=['ed25519:a_OHOQ'], min_valid=1622756012022)`
* `SynapseError: 401: Invalid signature for server pc.koesters.xyz:59003 with key ed25519:key1: Unable to verify signature for pc.koesters.xyz:59003: <class 'nacl.exceptions.BadSignatureError'> Signature was forged or corrupt`

Possibly only happens in worker setups?

### Version information

- **Homeserver**: fosdem.org (& others, all running with workers)
- **Version**: 1.35.0 & 1.35.1

### PR DESCRIPTION
**Pull request by erikjohnston**
Fixes #10123

Broke in #10082

### CODE REVIEW

File: synapse/handlers/federation.py [hunk 1]
```diff
@@ -2077,7 +2079,7 @@ async def _auth_and_persist_events(
         self,
         origin: str,
         room_id: str,
-        event_infos: Iterable[_NewEventInfo],
+        event_infos: Collection[_NewEventInfo],
```
Comment (anoadragon453) [inline synapse/handlers/federation.py:2082]: Is the length of `event_infos` necessary for this function? As far as I can see all usage only requires an Iterable.

File: synapse/util/linked_list.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
+# Copyright 2021 The Matrix.org Foundation C.I.C.
```
[the remark this answers is not in the document]
Comment (erikjohnston) [reply synapse/util/linked_list.py:1]: :facepalm:

File: synapse/handlers/federation.py (hunk 1 above)
Comment (erikjohnston) [reply synapse/handlers/federation.py:2082]: Yeah, iterables don't necessarily have a size functon.

Comment (erikjohnston) [reply synapse/handlers/federation.py:2082]: Oh, yes indeed.

Comment (anoadragon453) [reply synapse/handlers/federation.py:2082]: Huh, TIL!

**Review by anoadragon453 — APPROVED** [review]

File: synapse/handlers/federation.py (hunk 1 above)
Comment (erikjohnston) [reply synapse/handlers/federation.py:2082]: At least, I'm fairly sure iterables don't have a size and so you can't check for emptiness...

**Pull request merged**